pkg/cache/sql/db/client.go

```diff
@@ -0,0 +1,313 @@
+// and returns a slice of those objects.
+func (c *Client) ReadObjects(rows Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	// defer rows.Close()
+
+	// rt := time.Now()
+	var result []any
+	for rows.Next() {
+		data, err := c.decryptScan(rows, shouldDecrypt)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+		singleResult, err := fromBytes(data, typ)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+		result = append(result, singleResult.Elem().Interface())
+	}
+	err := rows.Err()
+	if err != nil {
+		return nil, closeRowsOnError(rows, err)
+	}
+
+	err = rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// ReadStrings scans the given rows into strings, and then returns the strings as a slice.
+func (c *Client) ReadStrings(rows Rows) ([]string, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+
+	var result []string
+	for rows.Next() {
+		var key string
+		err := rows.Scan(&key)
+		if err != nil {
+			return nil, closeRowsOnError(rows, err)
+		}
+
+		result = append(result, key)
+	}
+	err := rows.Err()
+	if err != nil {
+		return nil, closeRowsOnError(rows, err)
+	}
+
+	err = rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// Begin attempt to begin a transaction, and returns it along with a function for unlocking the
+// database once the transaction is done.
+func (c *Client) Begin() (TXClient, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	sqlTx, err := c.conn.Begin()
+	if err != nil {
+		return nil, err
+	}
+	return transaction.NewClient(sqlTx), nil
+}
+
+func (c *Client) decryptScan(rows Rows, shouldDecrypt bool) ([]byte, error) {
+	var data, dataNonce, eDEK, dekNonce sql.RawBytes
+	err := rows.Scan(&data, &dataNonce, &eDEK, &dekNonce)
+	if err != nil {
+		return nil, err
+	}
+	if c.decryptor != nil && shouldDecrypt {
+		decryptedData, err := c.decryptor.Decrypt(data, dataNonce, eDEK, dekNonce)
+		if err != nil {
+			return nil, err
+		}
+		return decryptedData, nil
+	}
+	return data, nil
+}
+
+// Upsert used to be called upsertEncrypted in store package before move
+func (c *Client) Upsert(tx TXClient, stmt *sql.Stmt, key string, obj any, shouldEncrypt bool) error {
+	objBytes := toBytes(obj)
+	// object, dek, deknonce, datanonce
+	var dataNonce, encryptedDEK, dekNonce []byte
+	var err error
+	if c.encryptor != nil && shouldEncrypt {
+		objBytes, dataNonce, encryptedDEK, dekNonce, err = c.encryptor.Encrypt(objBytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	return tx.StmtExec(tx.Stmt(stmt), key, objBytes, dataNonce, encryptedDEK, dekNonce)
+}
+
+// toBytes encodes an object to a byte slice
+func toBytes(obj any) []byte {
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err := enc.Encode(obj)
+	if err != nil {
+		panic(errors.Wrap(err, "Error while gobbing object"))
```

Comment:
Hmm, unsure. What should Rancher do with that error? Potentially important data intended to be saved is now lost, because it did not reach the DB.

Reading into gob's source, this method will return an error if a nil pointer is passed, or if a type points to itself in a cycle (eg. `T *T`), or if the type is not gobbable for some other reason. In all those cases the only possibility is a code fix, therefore failing fast seems like the best option to me, and restarting Rancher the least worse option at disposal.

What do you think?

Reply:
My concern here was that specially craft input could trigger the panic therefore crashing the entire application, potentially causing a DoS. On the next review I will take a deeper look at this.

Reply:
I do not disagree that may happen (though unlikely, as objects are expected to come from Kubernetes, where such conditions should not be possible).

What I mean is that, even in case it does happen, we would not have a much better alternative to bubble up the error all the way up to the main method and basically restart the application anyway.

I am resolving this for now, let's revisit when the new audit happens.